runtime/lua/vim/func/_memoize.lua

```diff
@@ -33,27 +33,54 @@ local function resolve_hash(hash)
   return hash
 end
 
+---@return table<any,table<any,any>>
+local function get_weak_table()
+  return setmetatable({}, { __mode = 'kv' })
+end
+
+--- @alias vim.func.Hash integer|string|fun(...): any
+
+--- @class vim.func.MemoizeTable
+--- @field clear fun(self, ...)
+
 --- @generic F: function
---- @param hash integer|string|fun(...): any
+--- @param hash vim.func.Hash
 --- @param fn F
 --- @return F
+--- @overload fun(hash: vim.func.Hash, fn: fun(...): any): vim.func.MemoizeTable
 return function(hash, fn)
   vim.validate({
```

Comment:
A better way of doing this would.be for `_memoize` to return 2 functions instead of one. That way you are not polluting the original function.